File: src/main/java/com/machfour/macros/ingredients/CompositeFoodAdapter.java
package com.machfour.macros.ingredients;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

class CompositeFoodAdapter extends TypeAdapter<CompositeFoodSpec> {
    // for reading the arrays of ingredients
    static final IngredientsAdapter ingredientsReader = new IngredientsAdapter();

    @Override
    public void write(JsonWriter writer, CompositeFoodSpec spec) throws IOException {
        if (spec == null) {
            writer.nullValue();
            return;
        }
        // TODO
        return;
    }

    @Override
    public CompositeFoodSpec read(JsonReader reader) throws IOException {
        String indexName = null;
        String variety = null;
        String name = null;
        String notes = null;
        List<IngredientSpec> ingredients = new ArrayList<>();
        reader.beginObject();
        while (reader.hasNext()) {
            String key = reader.nextName();
            switch (key) {
                case "index_name":
                    indexName = reader.nextString();
                    break;
                case "name":
                    name = reader.nextString();
                    break;
                case "variety":
                    variety = reader.nextString();
                    break;
                case "notes":
                    notes = reader.nextString();
                    break;
                case "ingredients":
                    reader.beginArray();
                    while (reader.hasNext()) {
                        ingredients.add(ingredientsReader.read(reader));
                    }
                    reader.endArray();
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.endObject();
        CompositeFoodSpec spec = new CompositeFoodSpec(indexName, name, variety, notes);
        spec.addIngredients(ingredients);
        return spec;
    }
}

File: src/main/kotlin/com/machfour/macros/ingredients/CompositeFoodAdapter.kt
package com.machfour.macros.ingredients

import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import java.io.IOException

internal class CompositeFoodAdapter : TypeAdapter<CompositeFoodSpec>() {
    @Throws(IOException::class)
    override fun write(writer: JsonWriter, spec: CompositeFoodSpec) {
        // TODO
        return
    }

    @Throws(IOException::class)
    override fun read(reader: JsonReader): CompositeFoodSpec {
        var indexName: String? = null
        var variety: String? = null
        var name: String? = null
        var notes: String? = null
        val ingredients: MutableList<IngredientSpec> = ArrayList()
        reader.beginObject()
        while (reader.hasNext()) {
            when (reader.nextName()) {
                "index_name" -> indexName = reader.nextString()
                "name" -> name = reader.nextString()
                "variety" -> variety = reader.nextString()
                "notes" -> notes = reader.nextString()
                "ingredients" -> {
                    reader.beginArray()
                    while (reader.hasNext()) {
                        ingredients.add(ingredientsReader.read(reader))
                    }
                    reader.endArray()
                }
                else -> reader.skipValue()
            }
        }
        reader.endObject()
        requireNotNull(indexName) { "Index name cannot be null" }
        requireNotNull(name) { "name cannot be null" }
        val spec = CompositeFoodSpec(indexName, name, variety, notes)
        spec.addIngredients(ingredients)
        return spec
    }

    companion object {
        // for reading the arrays of ingredients
        val ingredientsReader = IngredientsAdapter()
    }
}